torch/csrc/api/include/torch/detail/ordered_dict.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <ATen/Error.h>
+
+#include <cstdint>
+#include <functional>
+#include <initializer_list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace detail {
+
+/// A simple ordered dictionary implementation, akin to Python's `OrderedDict`.
+template <typename T>
+class OrderedDict {
+ public:
+  struct Item {
+    Item(std::string key_, T value_)
+        : key(std::move(key_)), value(std::move(value_)) {}
+
+    T& operator*() {
+      return value;
+    }
+    const T& operator*() const {
+      return value;
+    }
+    T* operator->() {
+      return &value;
+    }
+    const T* operator->() const {
+      return &value;
+    }
+
+    const std::string key;
+    T value;
+  };
+
+  using Iterator = typename std::vector<Item>::iterator;
+  using ConstIterator = typename std::vector<Item>::const_iterator;
+
+  OrderedDict() = default;
+
+  // Copy we have to do ourselves, because items' keys are const, so we have to
+  // re-insert the items.
+  OrderedDict(const OrderedDict& other) : index_(other.index_) {
+    for (const auto& item : other.items_) {
+      items_.push_back(item);
+    }
+  }
+
+  OrderedDict& operator=(const OrderedDict& other) {
+    index_ = other.index_;
+    items_.clear();
+    for (const auto& item : other.items_) {
+      items_.push_back(item);
+    }
+    return *this;
+  }
+
+  // Move works by default, because you can move-construct vectors of const
+  // values..
+  OrderedDict(OrderedDict&& other) = default;
+  OrderedDict& operator=(OrderedDict&& other) = default;
+
+  ~OrderedDict() = default;
+
+  /*implicit */ OrderedDict(std::initializer_list<Item> initializer_list) {
+    items_.reserve(initializer_list.size());
+    for (auto& item : initializer_list) {
+      // Copy the key here and move it into the index.
+      items_.emplace_back(item.key, std::move(item.value));
+      index_.emplace(std::move(item.key), size() - 1);
+    }
+  }
+
+  Iterator begin() {
+    return items_.begin();
+  }
+
+  ConstIterator begin() const {
+    return items_.begin();
+  }
+
+  Iterator end() {
+    return items_.end();
+  }
+
+  ConstIterator end() const {
+    return items_.end();
+  }
+
+  Item& front() {
+    return items_.front();
+  }
+
+  const Item& front() const {
+    return items_.front();
+  }
+
+  Item& back() {
+    return items_.back();
+  }
+
+  const Item& back() const {
+    return items_.back();
+  }
+
+  Item& operator[](size_t index) {
+    return items_[index];
+  }
+
+  const Item& operator[](size_t index) const {
+    return items_[index];
+  }
+
+  T& operator[](const std::string& key) {
+    return get(key);
+  }
+
+  const T& operator[](const std::string& key) const {
+    return get(key);
+  }
+
+  template <typename Key, typename Value>
+  T& insert(Key&& key, Value&& value) {
```

Comment:
Just take a `std::string` by value.

(I'm not sure why you put a universal reference here; maybe because you're trying to make it work with lvalues and rvalues? But see also https://stackoverflow.com/questions/21605579/how-true-is-want-speed-pass-by-value?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa )

Reply:
It's to perfect forward the key. This way if I pass a `const char*` I get one construction in `index_.emplace`. If I use `std::string` I get one construction *and one extra move*

Reply:
Did you mean the key or the value?

Reply:
The key. Do we really expect `insert` to be called in a sufficiently tight loop for the extra move to matter? Surely we will get soaked first by `unordered_map`.